new/resource-actions.md

```diff
@@ -0,0 +1,199 @@
+
+```
+* x[blah] action blah
+  * file[/Users/jkeiser/x.txt] action create (up to date)
+   (up to date)
+```
+
+We suggest not actually showing sub-resources as separate things in output:
+
+```
+* x[blah] action blah (up to date)
+```
+
+An "update" would look like this (again, removes the extra nesting):
+
+```
+* x[blah] action blah
+  - create new file /Users/jkeiser/x.txt
+  - update content in file /Users/jkeiser/x.txt from none to a591a6
+  --- /Users/jkeiser/x.txt	2015-04-29 08:15:14.000000000 -0700
+  +++ /Users/jkeiser/.x.txt20150429-56731-6xk0lx	2015-04-29 08:15:14.000000000 -0700
+  @@ -1 +1,2 @@
+  +Hello World
+```
+
+The principle is: the user doesn't see resources they don't type.
+
+### inline_recipe resource
+
+People need to support test-and-set and why-run in their resources. Given that
+actions are defined as recipes, the typical (and best) way to do that is
+resources (which have `not_if` and `only_if`, etc.). These resources work well,
+but sometimes you want to do a clump of actions based on a single `if`, or you
+want to nest them.
+
+`inline_recipe` is designed to do that job.  resource lets you write recipe
+code. It also has the parent in scope (so you can still access instance
+variables). The resources inside will compile immediately and be available in
+`inline_recipe.resources`.  The resources will converge when the `inline_recipe`
+converges.
+
+Like recipe actions, the resources inside the `inline_recipe` are not
+referenceable by `notifies` outside.  Unlike recipe actions, the resources show
+up in nested output (because the user typed them in the recipe).
+
+```ruby
+inline_recipe 'Create that one file' do
+  only_if { File.exist?('/please_create_that_file.txt') }
+
+  directory '/x' do
+  end
+  file '/x/y.txt' do
+  end
+end
+```
+
+`converge_by` (the current way of converging / updating) a resource will still
+be available for those who are used to it.
+
+### `Resource.why_run_description`
+
+When why run is enabled, we automatically prepend "Would" to the text of the
+why run description.  This doesn't always read well.  Herein we propose adding
+parameters for why-run safety to resources.
+
+The `converge_description` and `why_run_description` methods on a resource let
+you customize this if you want.  (By default, they will emit the same text as
+before.)
+
+```ruby
+ruby_block 'Delete /x.txt' do
+  why_run_description 'Would totally delete /x.txt'
+  only_if { File.exist?('/x.txt') }
+  action :run
+end
+```
+
+The `why_run_description` and `converge_description` can be lazy, and can thus
+be a calculated value run in lieu of the block.
+
+## Changes to existing things
+
+### Resource
+
+We add the following public API to `Resource`:
+
+- `self.action(action, class: nil, &block)`: creates a Provider with the given
+  action
+
+And modify the following:
+
+- `allowed_actions`: defaults to all actions in the resource's provider class(es).
+- `default_action`: defaults to `allowed_actions.first`.
```

Comment:
They do inherit properly, in this scheme at least.  Specifically, if parent has actions A, B and C, and the child adds D and E, allowed actions will be A, B and C, and default action will still be A (because it is first).